networkx/utils/backends.py

```diff
@@ -196,110 +182,172 @@ def _dispatch(
+        preserve_all_attrs=False,
+    ):
+        if func is None:
+            return functools.partial(
+                _dispatch,
+                name=name,
+                graphs=graphs,
+                edge_attrs=edge_attrs,
+                node_attrs=node_attrs,
+                preserve_edge_attrs=preserve_edge_attrs,
+                preserve_node_attrs=preserve_node_attrs,
+                preserve_graph_attrs=preserve_graph_attrs,
+                preserve_all_attrs=preserve_all_attrs,
+            )
+        if isinstance(func, str):
+            raise TypeError("'name' and 'graphs' must be passed by keyword") from None
+        # If name not provided, use the name of the function
+        if name is None:
+            name = func.__name__
+
+        self = object.__new__(cls)
+
+        # Make it look like the original function
+        functools.update_wrapper(self, func)
+        self.__defaults__ = func.__defaults__
+        self.__kwdefaults__ = func.__kwdefaults__
+
+        self.orig_func = func
+        self.name = name
+        self.edge_attrs = edge_attrs
+        self.node_attrs = node_attrs
+        self.preserve_edge_attrs = preserve_edge_attrs or preserve_all_attrs
+        self.preserve_node_attrs = preserve_node_attrs or preserve_all_attrs
+        self.preserve_graph_attrs = preserve_graph_attrs or preserve_all_attrs
+
+        if isinstance(graphs, str):
+            graphs = {graphs: 0}
+        elif len(graphs) == 0:
+            raise KeyError("'graphs' must contain at least one variable name") from None
+
+        # This dict comprehension is complicated for better performance; equivalent shown below.
+        self.optional_graphs = set()
+        self.list_graphs = set()
+        self.graphs = {
+            self.optional_graphs.add(val := k[:-1]) or val
+            if (last := k[-1]) == "?"
+            else self.list_graphs.add(val := k[1:-1]) or val
+            if last == "]"
+            else k: v
+            for k, v in graphs.items()
+        }
+        # The above is equivalent to:
+        # self.optional_graphs = {k[:-1] for k in graphs if k[-1] == "?"}
+        # self.list_graphs = {k[1:-1] for k in graphs if k[-1] == "]"}
+        # self.graphs = {k[:-1] if k[-1] == "?" else k: v for k, v in graphs.items()}
+
+        sig = inspect.signature(func)
+        if not any(
+            p.kind == inspect.Parameter.VAR_KEYWORD for p in sig.parameters.values()
+        ):
+            sig = sig.replace(
+                parameters=[
+                    *sig.parameters.values(),
+                    inspect.Parameter("backend_kwargs", inspect.Parameter.VAR_KEYWORD),
+                ]
+            )
+        self.__signature__ = sig
+
+        if name in _registered_algorithms:
+            raise KeyError(f"Algorithm already exists in dispatch registry: {name}")
+        _registered_algorithms[name] = self
+        return self
+
+    def __call__(self, *args, **kwargs):
```

Comment:
I was thinking if it makes sense to have a short circuit here if there are no new backends installed (except nx-loopback)? something like:
``` python
        if len(plugins.items) == 1:
            return self.orig_func(*args, **kwargs)
```
This will not fail elegantly if a backend graph is passed as an argument without installing the backend but we will have an early return without going through all the logic below this.

Reply:
I like the idea, but how would we test with the `nx-loopback` backend?

Reply:
Done. I added a fast path in the latest commit. We don't include `"nx-loopback"` as a backend unless it's being tested.